src/plugins/timeseries-plugin.ts

```diff
@@ -0,0 +1,259 @@
+        super();
+
+        this.styles = { ...defaultCPUPluginStyles, ...(styles ?? {}) };
+        this.data = [];
+        this.name = name;
+
+        this.height = 100;
+
+        this.setData(data);
+    }
+
+    override init(renderEngine: OffscreenRenderEngine, interactionsEngine: SeparatedInteractionsEngine) {
+        super.init(renderEngine, interactionsEngine);
+
+        this.interactionsEngine.on('change-position', this.handlePositionChange.bind(this));
+        this.interactionsEngine.on('hover', this.handleHover.bind(this));
+
+        this.interactionsEngine.on('up', this.handleMouseUp.bind(this));
+    }
+
+    handlePositionChange(position: { deltaX: number }) {
+        const startPositionX = this.renderEngine.parent.positionX;
+
+        this.interactionsEngine.setCursor('grabbing');
+
+        this.renderEngine.tryToChangePosition(position.deltaX);
+
+        if (startPositionX !== this.renderEngine.parent.positionX) {
+            this.renderEngine.parent.render();
+        }
+    }
+
+    handleMouseUp() {
+        this.interactionsEngine.clearCursor();
+    }
+
+    setData(data: TimeseriesPoint[]) {
+        this.data = data;
+
+        let min = Number.MAX_VALUE;
+        let max = Number.MIN_VALUE;
+        let first = Number.MAX_VALUE;
+        let last = Number.MIN_VALUE;
+
+        this.data.forEach(([ts, v]) => {
+            if (v < min) {
+                min = v;
+            }
+            if (v > max) {
+                max = v;
+            }
+
+            if (ts < first) {
+                first = ts;
+            }
+            if (ts > last) {
+                last = ts;
+            }
+        });
+
+        this.summary = {
+            min,
+            max,
+            first,
+            last,
+        };
+    }
+
+    calcRect(start: number, duration: number, isEnd: boolean) {
+        const w = duration * this.renderEngine.zoom;
+
+        return {
+            x: this.renderEngine.timeToPosition(start),
+            w: isEnd ? (w <= 0.1 ? 0.1 : w >= 3 ? w - 1 : w - w / 3) : w,
+        };
+    }
+
+    handleHover(region: HitRegion<number> | null) {
+        this.hoveredRegion = region;
+    }
+
+    override renderTooltip() {
+        if (this.hoveredRegion) {
+            const round = (v) => Math.round(v * 100) / 100;
+            const data = this.hoveredRegion.data as TimeseriesPointWithIndex;
+
+            this.renderEngine.renderTooltipFromData(
+                [
+                    { text: `Value: ${round(data[TimeseriesPointWithIndexX.value])}` },
+                    {
+                        text: `Timestamp: ${round(data[TimeseriesPointWithIndexX.ts])}ms`,
+                    },
+                ],
+                this.interactionsEngine.getGlobalMouse()
+            );
+            return true;
+        }
+        return false;
+    }
+
+    override render() {
+        if (this.data?.length === 0 || !this.summary) {
+            console.log('[render] no data - returned');
```

Comment:
not sure if we should log stuff to the user browser and create noise
if we should, then `[render] no data - returned` is not descriptive enough to figure out on what is the issue

Reply:
will happily remove the console.

Reply:
done